lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp

```diff
@@ -3337,24 +3339,44 @@ static void ReplaceMemcpy(Value *V, Value *Src, MemCpyInst *MC) {
 
       BitCastInst *DestBCI = cast<BitCastInst>(DestVal);
       BitCastInst *SrcBCI = cast<BitCastInst>(SrcVal);
-      if (!ArePointersToStructsOfIdenticalLayouts(DestBCI->getSrcTy(), SrcBCI->getSrcTy())) {
-        DXASSERT(0, "Can't handle structs of different layouts");
-        return;
-      }
 
-      const DataLayout &DL = SrcBCI->getModule()->getDataLayout();
-      unsigned SrcSize = DL.getTypeAllocSize(SrcBCI->getOperand(0)->getType()->getPointerElementType());
-      unsigned MemcpySize = cast<ConstantInt>(MC->getLength())->getZExtValue() * MC->getAlignment();
-      if (SrcSize != MemcpySize) {
-        DXASSERT(0, "Cannot handle partial memcpy");
-        return;
-      }
+      Type* DstTy = DestBCI->getSrcTy();
+      Type *SrcTy = SrcBCI->getSrcTy();
+      if (ArePointersToStructsOfIdenticalLayouts(DstTy, SrcTy)) {
+        const DataLayout &DL = SrcBCI->getModule()->getDataLayout();
+        unsigned SrcSize = DL.getTypeAllocSize(
+            SrcBCI->getOperand(0)->getType()->getPointerElementType());
+        unsigned MemcpySize =
+            cast<ConstantInt>(MC->getLength())->getZExtValue() *
+            MC->getAlignment();
```

Comment:
This multiplication by alingment looks bogus. I think it works because we put 1 in there.